### ISSUE TITLE
(Artillery) Loki Subsystem or Skill Changes cause Error window

### ISSUE THREAD

**Issue by PaladinOne** [pyfa-org/Pyfa#1989]
I've been playing around with several Loki fits in recent days and I've encountered a rather irritating issue: replacing a Subsystem (any Subsystem) on a Loki, fitted with one or more of any Medium Artillery turret, triggers an Error window. Furthermore, after having triggered this bug report at least once, changing what Character is selected (for skills) also triggers the exact same report.

```
## Bug Report

pyfa v2.9.4dev3
EVE Data Version: 1503340 (2019-05-09 06:14:26)

OS version: Windows-10-10.0.17134-SP0
Python version: 3.6.8 (tags/v3.6.8:3c6b436a57, Dec 23 2018, 23:31:17) [MSC v.1916 32 bit (Intel)]
wxPython version: 4.0.4 (wxWidgets 3.0.5)
SQLAlchemy version: 1.3.3
Logbook version: 1.4.3
Requests version: 2.21.0
Dateutil version: 2.8.0

####################

Traceback (most recent call last):
  File "C:\projects\pyfa\gui\statsPane.py", line 75, in fitChanged
  File "C:\projects\pyfa\gui\builtinStatsViews\firepowerViewFull.py", line 190, in refreshPanel
  File "C:\projects\pyfa\gui\builtinStatsViews\firepowerViewFull.py", line 164, in <lambda>
  File "C:\projects\pyfa\eos\saveddata\fit.py", line 279, in getWeaponDps
  File "C:\projects\pyfa\eos\saveddata\fit.py", line 1486, in calculateWeaponDmgStats
  File "C:\projects\pyfa\eos\saveddata\module.py", line 460, in getDps
  File "C:\projects\pyfa\eos\saveddata\module.py", line 858, in getCycleParameters
AttributeError: 'NoneType' object has no attribute 'factorReload'
```

Example fit:
```
[Loki, Active Rat]

[Empty Low slot]
[Empty Low slot]
[Empty Low slot]
[Empty Low slot]

[Empty Med slot]
[Empty Med slot]
[Empty Med slot]
[Empty Med slot]
[Empty Med slot]
[Empty Med slot]

720mm Howitzer Artillery II, EMP M
720mm Howitzer Artillery II, EMP M
720mm Howitzer Artillery II, EMP M
720mm Howitzer Artillery II, EMP M
720mm Howitzer Artillery II, EMP M
[Empty High slot]
[Empty High slot]

[Empty Rig slot]
[Empty Rig slot]
[Empty Rig slot]

Loki Core - Dissolution Sequencer
Loki Offensive - Projectile Scoping Array
Loki Propulsion - Wake Limiter
Loki Defensive - Adaptive Defense Node
```

Replacing any Subsystem (either via right-click context or via doubleclicking from sidebar) causes the Error.
This Error is not present on Loki fits that do not have at least one Artillery (650 or 720, any variant) fitted.

However, in the midst of typing this report (having tested various different turrets and isolated it to only be on Artillery, not autocannons or missiles), I made a completely fresh Proteus fit, and equipping it with its first Subsystem caused the Error again.
If I didn't know better I'd say the bug was contagious and infecting my other Strat Cruisers.

**Comment by PaladinOne** [comment]
Attempting to Delete my new Proteus fit (with a single Subsystem) failed to delete it and in stead produced another error

```
####################

Traceback (most recent call last):
  File "C:\Python36\lib\site-packages\sqlalchemy\engine\base.py", line 1244, in _execute_context
  File "C:\Python36\lib\site-packages\sqlalchemy\engine\default.py", line 552, in do_execute
sqlite3.IntegrityError: NOT NULL constraint failed: modules.fitID

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\projects\pyfa\eos\db\saveddata\queries.py", line 551, in commit
  File "C:\Python36\lib\site-packages\sqlalchemy\orm\session.py", line 1026, in commit
  File "C:\Python36\lib\site-packages\sqlalchemy\orm\session.py", line 493, in commit
  File "C:\Python36\lib\site-packages\sqlalchemy\orm\session.py", line 472, in _prepare_impl
  File "C:\Python36\lib\site-packages\sqlalchemy\orm\session.py", line 2451, in flush
  File "C:\Python36\lib\site-packages\sqlalchemy\orm\session.py", line 2589, in _flush
  File "C:\Python36\lib\site-packages\sqlalchemy\util\langhelpers.py", line 68, in __exit__
  File "C:\Python36\lib\site-packages\sqlalchemy\util\compat.py", line 129, in reraise
  File "C:\Python36\lib\site-packages\sqlalchemy\orm\session.py", line 2549, in _flush
  File "C:\Python36\lib\site-packages\sqlalchemy\orm\unitofwork.py", line 422, in execute
  File "C:\Python36\lib\site-packages\sqlalchemy\orm\unitofwork.py", line 589, in execute
  File "C:\Python36\lib\site-packages\sqlalchemy\orm\persistence.py", line 245, in save_obj
  File "C:\Python36\lib\site-packages\sqlalchemy\orm\persistence.py", line 1120, in _emit_insert_statements
  File "C:\Python36\lib\site-packages\sqlalchemy\engine\base.py", line 988, in execute
  File "C:\Python36\lib\site-packages\sqlalchemy\sql\elements.py", line 287, in _execute_on_connection
  File "C:\Python36\lib\site-packages\sqlalchemy\engine\base.py", line 1107, in _execute_clauseelement
  File "C:\Python36\lib\site-packages\sqlalchemy\engine\base.py", line 1248, in _execute_context
  File "C:\Python36\lib\site-packages\sqlalchemy\engine\base.py", line 1466, in _handle_dbapi_exception
  File "C:\Python36\lib\site-packages\sqlalchemy\util\compat.py", line 383, in raise_from_cause
  File "C:\Python36\lib\site-packages\sqlalchemy\util\compat.py", line 128, in reraise
  File "C:\Python36\lib\site-packages\sqlalchemy\engine\base.py", line 1244, in _execute_context
  File "C:\Python36\lib\site-packages\sqlalchemy\engine\default.py", line 552, in do_execute
sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) NOT NULL constraint failed: modules.fitID
[SQL: INSERT INTO modules ("fitID", "itemID", "baseItemID", "mutaplasmidID", "dummySlot", "chargeID", state, projected, position, created, modified, "spoolType", "spoolAmount") VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)]
[parameters: (None, None, None, None, 1, None, <FittingModuleState.ONLINE: 0>, 0, 7, '2019-05-27 03:09:12.205797', None, None, None)]
(Background on this error at: http://sqlalche.me/e/gkpj)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\projects\pyfa\gui\builtinShipBrowser\sfBrowserItem.py", line 339, in OnLeftUp
  File "C:\projects\pyfa\gui\builtinShipBrowser\sfBrowserItem.py", line 51, in DoCallback
  File "C:\projects\pyfa\gui\builtinShipBrowser\fitItem.py", line 363, in deleteBtnCB
  File "C:\projects\pyfa\gui\builtinShipBrowser\fitItem.py", line 381, in deleteFit
  File "C:\projects\pyfa\service\fit.py", line 210, in deleteFit
  File "C:\projects\pyfa\eos\db\saveddata\queries.py", line 545, in remove
  File "C:\projects\pyfa\eos\db\saveddata\queries.py", line 555, in commit
TypeError: __init__() missing 2 required positional arguments: 'params' and 'orig'
```

**Comment by DarkFenX** [comment]
Your first issue is old pyfa issue of relying on attribute links handled by SQLAlchemy without doing flush()/commit() before that. Latest build with graphs had some changes regarding how module cycles and volleys are calculated, which made this bug to pop up again.

Should be fixed in bc84c20cb235e4694bf0a2fe79f412e196d0c9b3. At least it doesn't occur for me anymore.

And hey, thanks for the report!
